src/ConnectedMode.UnitTests/BoundSolutionUpdateHandlerTests.cs

```diff
@@ -0,0 +1,78 @@
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using SonarLint.VisualStudio.ConnectedMode.Suppressions;
+using SonarLint.VisualStudio.Core.Binding;
+using SonarLint.VisualStudio.TestInfrastructure;
+
+namespace SonarLint.VisualStudio.ConnectedMode.UnitTests
+{
+    [TestClass]
+    public class BoundSolutionUpdateHandlerTests
+    {
+        [TestMethod]
+        public void MefCtor_CheckIsExported()
+        {
+            MefTestHelpers.CheckTypeCanBeImported<BoundSolutionUpdateHandler, BoundSolutionUpdateHandler>(
+                MefTestHelpers.CreateExport<IActiveSolutionBoundTracker>(),
+                MefTestHelpers.CreateExport<ISuppressionIssueStoreUpdater>());
+        }
+
+        [TestMethod]
+        public void Ctor_SubscribesToEvents()
+        {
+            var activeSolutionTracker = new Mock<IActiveSolutionBoundTracker>();
+
+            _ = new BoundSolutionUpdateHandler(activeSolutionTracker.Object, Mock.Of<ISuppressionIssueStoreUpdater>());
+
+            activeSolutionTracker.VerifyAdd(x => x.SolutionBindingChanged += It.IsAny<EventHandler<ActiveSolutionBindingEventArgs>>(), Times.Once());
+            activeSolutionTracker.VerifyAdd(x => x.SolutionBindingUpdated += It.IsAny<EventHandler>(), Times.Once());
+        }
+
+        [TestMethod]
+        public void InvokeEvents_SuppressionIssueStoreUpdaterIsCalled()
+        {
+            var activeSolutionTracker = new Mock<IActiveSolutionBoundTracker>();
+            var suppressionIssueStoreUpdater = new Mock<ISuppressionIssueStoreUpdater>();
+
+            _ = new BoundSolutionUpdateHandler(activeSolutionTracker.Object, suppressionIssueStoreUpdater.Object);
+
+            activeSolutionTracker.Raise(x => x.SolutionBindingChanged += null, new ActiveSolutionBindingEventArgs(BindingConfiguration.Standalone));
+            suppressionIssueStoreUpdater.Verify(x => x.UpdateAllServerSuppressionsAsync(), Times.Once);
+
+            activeSolutionTracker.Raise(x => x.SolutionBindingUpdated += null, EventArgs.Empty);
+            suppressionIssueStoreUpdater.Verify(x => x.UpdateAllServerSuppressionsAsync(), Times.Exactly(2));
+        }
+
+        [TestMethod]
+        public void Dispose_UnsubscribesToEvent()
```

Comment:
Wouldit be a thing to test that if I raise an event after unsubscribing it doesnt call my method anymore? 

ie. 

Dispose()
activeSolutionTracker.BindingChanged.Raise()
suppressionIssueStoreUpdater.Verify((x => x.UpdateAllServerSuppressionsAsync(), Times.Never)

Reply:
I think we have a couple of tests that do that, although I think it's unnecessary here since you are calling `VerifyRemove` to check that the event handler is unregistered.

Essentially, one is white-box test (i.e. the test has some knowledge of how the method is implemented), and the other is a black-box test (i.e. the test knows nothing about how the test is implemented).

Yours is a white-box version - it checks that the product code unregisters a specific event handler.

The alternative code you suggested above is a black-box version: it doesn't know what the product code does internally, but it observes the externally visible effects (i.e. whether `UpdateAllServerSuppressionsAsync` is called or not).